Add specs for etc-local and stack.

kierenevans opened the pull request:
Add specs for the existing behaviour of `magento-ng::stack` and `magento-ng::etc-local`.

The etc-local recipe has a load of functionality hidden in the template that it writes out. Any advice on splitting up that mammoth spec is appreciated!

I have had to modify the `magento-local.xml.erb` template as I couldn't get a definition of memcached server details in symbol format to work.

I was hoping to have the following in the etc-local spec too, with appropriate stubbing of ![secure_password](https://github.com/opscode-cookbooks/openssl/blob/v2.0.0/libraries/secure_password.rb] to test out the generation of the crypt key but unfortunately I ran into https://github.com/edelight/chef-solo-search/issues/74 :
```context 'with chef server' do
  let(:chef_run) do
    ChefSpec::ServerRunner.new.converge(described_recipe)
  end

  it 'should not raise an error if crypt key is not specifed' do
    expect { chef_run }.to_not raise_error
  end

  it_should_behave_like 'etc-local'
end```

andytson-inviqa commented:
Nice work, will review it hopefully next week.

Regarding the many behaviours in the recipe, it'd be possible to stick the configs into different xml files and receipes (tho not what we usually do in Magento).

they'd need to be merged in after enterprise.xml

As for chef-solo-search, the dependency was meant to be for `config-driven-helper::capistrano` deploy user creation, however like your Github issue, it was dependant on a release of the new version of that cookbook as well to fix an issue with it not supporting encrypted data bags (and therefore the users cookbook). I replaced it with custom implementation which does, till that is done, however left the metadata dependency in hoping it wouldn't take too long before I could convert it back.

Perhaps we could drop the users cookbook and chef-solo-search from config-driven-helper for now? Would that solve testing here?

kierenevans commented:
Thanks :)

If we'd prefer to keep it a big config file, I wonder if ![partial templates](https://docs.chef.io/resource_template.html#partial-templates) would be of help.

It might be worth dropping the users and chef-solo-search cookbooks from config-driven-helper then and reviewing the situation when a new version is pushed up. The alternative is switching to the git version of chef-solo-search.

I went ahead and added another spec for when running on chef server, with a workaround to avoid the chef-solo-search issue. I added a blank cookbook with the right metadata and pointed berkshelf at that (see https://github.com/kierenevans/chef-magento-ng/commit/7e5669d3b4c87986150d1a09910efdb26151bac5 ).

A few issues were encountered with the server run so I needed to make a few changes to the etc-local recipe. The issues were:
1. secure_password was an unknown method due to openssl's library not being included. The message chef threw was `expected no Exception, got #<NameError: No resource, method, or local variable named `secure_password' for `Chef::Recipe "etc-local"'>`
2. No crypt key was written to the local.xml file due to the `hash_only_merge` that happens beforehand. This does a `.dup` on both arguments to the merge.

andytson commented:
Sorry, the partial templates link gives a 404.

Btw, we've released config-driven-helper 1.4.3 now (change in tag structure as well to v1.4.3). This has the users and chef-solo-search dependencies removed

kierenevans commented:
I've rebased on top of master and tidied up a bit :)
